### ISSUE TITLE
CPU and RAM hog when compiling infinitely-sized enums

### ISSUE THREAD

**Issue by nabijaczleweli** [rust-lang/rust#33557]
rustc seems to hog memory and CPU at an alarming rate when faced with an infinitely-sized `enum`

I tried this code:

```rust
enum Expression {
    Sum { augend: Expression, addend: Expression },
    Difference { minuend: Expression, subtrahend: Expression },
    Product { multiplier: Expression, multiplicand: Expression },
    Quotient { numerator: Expression, denomenator: Expression },
    Power { base: Expression, exponent: Expression }
}

fn main() {}
```

I expected to see this happen: compiler dies and spits out a nice error message.

Instead, this happened:
After only a couple of seconds I saw this: ![procexp SS](http://i.imgur.com/QGIb3bL.png) (CPU usage is per-hyperthread, so that's actually 100% of one). The compiler responds to Ctrl+C.

## Meta

`rustc --version --verbose`:
```
rustc 1.8.0 (db2939409 2016-04-11)
binary: rustc
commit-hash: db2939409db26ab4904372c82492cd3488e4c44e
commit-date: 2016-04-11
host: x86_64-pc-windows-gnu
release: 1.8.0
```
OS: Win10

Backtrace: N/A

**Comment by oli-obk** [comment]
This is already fixed in 1.9 beta

**Comment by nabijaczleweli** [comment]
Alright, thanks
